### PR TITLE
[Merged by Bors] - feat(GroupTheory/Congruence/*): add some missing lemma and constructions of group congruence

### PR DESCRIPTION
**Pull request by jjaassoonn**
1. order isomorphism between congruence on group and opposite groups
2. congruence relation is preserved by finite product/sum

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/GroupTheory/Congruence/BigOperators.lean [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+@[to_additive "Additive congruence relations preserve finite sum."]
+protected theorem finset_prod {ι M : Type*} [CommMonoid M] (c : Con M) (s : Finset ι)
+    {f g : ι → M} (h : ∀ i ∈ s, c (f i) (g i)) :
+    c (s.prod f) (s.prod g) := by
+  classical
+  induction s using Finset.cons_induction_on with
+  | h₁ => simpa using c.refl 1
+  | @h₂ i s hi ih =>
+    rw [Finset.prod_cons hi, Finset.prod_cons hi]
+    exact c.mul (h _ (by simp)) <| ih fun i hi ↦ h _ (by aesop)
```
Comment (alreadydone) [inline Mathlib/GroupTheory/Congruence/BigOperators.lean:59]:
```suggestion
    exact c.mul (h _ <| .head _) <| ih fun k hk ↦ h _ (.tail _ hk)

/-- Multiplicative congruence relations preserve product indexed by a multiset. -/
@[to_additive "Additive congruence relations preserve sum indexed by a multiset."]
protected theorem multiset_prod {ι M : Type*} [CommMonoid M] (c : Con M) {s : Multiset ι}
    {f g : ι → M} (h : ∀ x ∈ s, c (f x) (g x)) :
    c (s.map f).prod (s.map g).prod := by
  rcases s; simpa using c.list_prod h

/-- Multiplicative congruence relations preserve finite product. -/
@[to_additive "Additive congruence relations preserve finite sum."]
protected theorem finset_prod {ι M : Type*} [CommMonoid M] (c : Con M) (s : Finset ι)
    {f g : ι → M} (h : ∀ i ∈ s, c (f i) (g i)) :
    c (s.prod f) (s.prod g) :=
  c.multiset_prod h
```

**Review by alreadydone — APPROVED** [review]

File: Mathlib/GroupTheory/Congruence/BigOperators.lean [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+/-- Multiplicative congruence relations preserve product indexed by a multiset. -/
+@[to_additive "Additive congruence relations preserve sum indexed by a multiset."]
+protected theorem multiset_prod {ι M : Type*} [CommMonoid M] (c : Con M) {s : Multiset ι}
+    {f g : ι → M} (h : ∀ x ∈ s, c (f x) (g x)) :
+    c (s.map f).prod (s.map g).prod := by
+  rcases s; simpa using c.list_prod h
+
+/-- Multiplicative congruence relations preserve finite product. -/
+@[to_additive "Additive congruence relations preserve finite sum."]
+protected theorem finset_prod {ι M : Type*} [CommMonoid M] (c : Con M) (s : Finset ι)
```
Comment (YaelDillies) [inline Mathlib/GroupTheory/Congruence/BigOperators.lean:42]:
```suggestion
protected theorem finsetProd {ι M : Type*} [CommMonoid M] (c : Con M) (s : Finset ι)
```

File: Mathlib/GroupTheory/Congruence/BigOperators.lean [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+  induction l with
+  | nil =>
+    simpa only [List.map_nil, List.prod_nil] using c.refl 1
+  | cons x xs ih =>
+    rw [List.map_cons, List.map_cons, List.prod_cons, List.prod_cons]
+    exact c.mul (h _ <| .head _) <| ih fun k hk ↦ h _ (.tail _ hk)
+
+/-- Multiplicative congruence relations preserve product indexed by a multiset. -/
+@[to_additive "Additive congruence relations preserve sum indexed by a multiset."]
+protected theorem multiset_prod {ι M : Type*} [CommMonoid M] (c : Con M) {s : Multiset ι}
```
Comment (YaelDillies) [inline Mathlib/GroupTheory/Congruence/BigOperators.lean:35]:
```suggestion
protected theorem multisetProd {ι M : Type*} [CommMonoid M] (c : Con M) {s : Multiset ι}
```

File: Mathlib/GroupTheory/Congruence/BigOperators.lean [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+
+-/
+
+open BigOperators
+
+namespace Con
+
+/-- Multiplicative congruence relations preserve product indexed by a list. -/
+@[to_additive "Additive congruence relations preserve sum indexed by a list."]
+protected theorem list_prod {ι M : Type*} [CommMonoid M] (c : Con M) {l : List ι} {f g : ι → M}
```
Comment (YaelDillies) [inline Mathlib/GroupTheory/Congruence/BigOperators.lean:23]:
```suggestion
protected theorem listProd {ι M : Type*} [CommMonoid M] (c : Con M) {l : List ι} {f g : ι → M}
```

**Review by jcommelin — COMMENTED** [review]
Thanks :tada:

bors merge

File: Mathlib/GroupTheory/Congruence/Opposite.lean [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+variable {M : Type*} [Mul M]
+
+namespace Con
+
+/-- If `c` is a multiplicative congruence on `M`, then `(a, b) ↦ c b.unop a.unop` is a
+multiplicative congruence on `Mᵐᵒᵖ`-/
+@[to_additive "If `c` is an additive congruence on `M`, then `(a, b) ↦ c b.unop a.unop` is an
+additive congruence on `Mᵃᵒᵖ`"]
+def op (c : Con M) : Con Mᵐᵒᵖ where
+  r a b := c b.unop a.unop
```
Comment (eric-wieser) [inline Mathlib/GroupTheory/Congruence/Opposite.lean:27]: Why are these swapped, and not in the same order?

**Pull request closed without merging**